contracts/zero-ex/contracts/src/features/ERC721OrdersFeature.sol

```diff
@@ -0,0 +1,1112 @@
+        } else {
+            // Transfer the ERC20 token from the buyer to the seller.
+            _transferERC20TokensFrom(
+                order.erc20Token, 
+                order.maker, 
+                taker, 
+                order.erc20TokenAmount
+            );
+        }
+
+        if (takerCallbackData.length > 0) {
+            require(
+                taker != address(this), 
+                "ERC721OrdersFeature::_sellERC721/CANNOT_CALLBACK_SELF"
+            );
+            // Invoke the callback
+            bytes4 callbackResult = IERC721OrderCallback(taker)
+                .zeroExERC721OrderCallback(takerCallbackData);
+            // Check for the magic success bytes
+            require(
+                callbackResult == TAKER_CALLBACK_MAGIC_BYTES,
+                "ERC721OrdersFeature::_sellERC721/CALLBACK_FAILED"
+            );            
+        }
+
+        // Transfer the ERC721 asset to the buyer.
+        // If this function is called from the 
+        // `onERC721Received` callback the Exchange Proxy
+        // holds the asset. Otherwise, transfer it from 
+        // the seller.
+        _transferERC721AssetFrom(
+            order.erc721Token,
+            isCallback ? address(this) : taker,
+            order.maker,
+            erc721TokenId
+        );
+
+        // The buyer pays the order fees.
+        _payFees(order, order.maker, false);
+
+        emit ERC721OrderFilled(
+            order.direction,
+            order.erc20Token,
+            order.erc20TokenAmount,
+            order.erc721Token,
+            erc721TokenId,
+            order.maker,
+            taker,
+            order.nonce
+        );
+    }
+
+    // Core settlement logic for buying an ERC721 asset.
+    // Used by `buyERC721` and `batchBuyERC721s`.
+    function _buyERC721(
+        LibERC721Order.ERC721Order memory order,
+        LibSignature.Signature memory signature,
+        uint256 ethAvailable,
+        bytes memory takerCallbackData
+    )
+        public
+        payable
+        returns (uint256 ethSpent)
+    {
+        // Check that the order can be filled.
+        _validateSellOrder(order, signature);
+
+        // Mark the order as filled.
+        _setOrderStatusBit(order.maker, order.nonce);
+
+        // Transfer the ERC721 asset to the buyer (`msg.sender`).
+        _transferERC721AssetFrom(
+            order.erc721Token,
+            order.maker,
+            msg.sender,
+            order.erc721TokenId
+        );
+
+        if (takerCallbackData.length > 0) {
+            require(
+                msg.sender != address(this), 
+                "ERC721OrdersFeature::_buyERC721/CANNOT_CALLBACK_SELF"
+            );
+            uint256 ethBalanceBeforeCallback = address(this).balance;
+            // Invoke the callback
+            bytes4 callbackResult = IERC721OrderCallback(msg.sender)
+                .zeroExERC721OrderCallback(takerCallbackData);
```

Comment:
Just calling out that we need to be mindful about this pattern in the future. If these contracts rely on checking `msg.sender == EP` to do something that costs someone money, it's a potential attack vector against them if we somehow let an arbitrary call to one of these addresses slide through. We've potentially expanded our liability here.